File: 프로그래머스/3/87391. 공 이동 시뮬레이션/공 이동 시뮬레이션.py
def move(dist, rng, boundary):
    rng[0] += 0 if (rng[0] == 0 and dist > 0) else dist
    rng[1] += 0 if (rng[1] == boundary - 1 and dist < 0) else dist
    
    if rng[0] >= boundary or rng[1] < 0:
        return False
    
    rng[0] = max(rng[0], 0)
    rng[1] = min(rng[1], boundary - 1)
    
    return True

def solution(n, m, x, y, queries):
    reverse_dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]
    
    row_range = [x, x]
    col_range = [y, y]
    
    for query in reversed(queries):
        if query[0] > 1: # move row
            dist = reverse_dir[query[0]][0] * query[1]
            if not move(dist, row_range, n):
                return 0
        
        else: # move column
            dist = reverse_dir[query[0]][1] * query[1]
            if not move(dist, col_range, m):
                return 0
    
    return (row_range[1] - row_range[0] + 1) * (col_range[1] - col_range[0] + 1)

File: 프로그래머스/3/87391. 공 이동 시뮬레이션/공 이동 시뮬레이션.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> reverse_dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

bool move(int dist, int& start, int& end, int range){
    start += (start == 0 && dist > 0) ? 0 : dist;
    end += (end == range - 1 && dist < 0) ? 0 : dist;
    
    if(start >= range || end < 0) return false;
    
    if(start < 0) start = 0;
    if(end >= range) end = range - 1;
    
    return true;
}

long long solution(int n, int m, int x, int y, vector<vector<int>> queries) {
    int row_start = x, row_end = x;
    int col_start = y, col_end = y;
    
    for(int i = queries.size() - 1; i >= 0; i--){
        if(queries[i][0] > 1){ // move row
            if(!move(reverse_dir[queries[i][0]][0] * queries[i][1], row_start, row_end, n)) return 0;
        }
        else{ // move column
            if(!move(reverse_dir[queries[i][0]][1] * queries[i][1], col_start, col_end, m)) return 0;
        }
    }
    return (long long)(row_end - row_start + 1) * (col_end - col_start + 1);
}